c2rust-analyze/src/rewrite/expr/distribute.rs

```diff
@@ -0,0 +1,107 @@
+use crate::rewrite::expr::mir_op::{self, MirRewrite};
+use crate::rewrite::expr::unlower::{MirOrigin, MirOriginDesc, PreciseLoc};
+use log::*;
+use rustc_hir::HirId;
+use rustc_middle::mir::Location;
+use rustc_middle::ty::TyCtxt;
+use std::collections::{BTreeMap, HashMap};
+
+struct RewriteInfo {
+    rw: mir_op::RewriteKind,
+    loc: PreciseLoc,
+    desc: MirOriginDesc,
+}
+
+/// Distributes MIR rewrites to HIR nodes.  This takes a list of MIR rewrites (from `mir_op`) and a
+/// map from MIR location to `HirId` (from `unlower`) and produces a map from `HirId` to a list of
+/// MIR rewrites.
+///
+/// Using the example from `unlower`:
+///
+/// ```text
+/// bb0[5]: Terminator { source_info: ..., kind: _4 = f(move _5) -> [return: bb1, unwind: bb2] }
+///   []: StoreIntoLocal, `f(y)`
+///   [Rvalue]: Expr, `f(y)`
+///   [Rvalue, CallArg(0)]: Expr, `y`
+/// ```
+///
+/// A MIR rewrite on `bb0[5]` `[Rvalue, CallArg(0)]` would be attached to the MIR
+/// `Expr` `y`, and a rewrite on `bb0[5]` `[Rvalue]` would be attached to `f(y)`.
+/// A MIR rewrite on `bb0[5]` `[]` (i.e. on the call terminator itself) would
```

Comment:
Is this the case with all terminators, and only terminators? For a statement that's not a terminator like `*x = 2` where `x` has `CELL` permissions I would expect an `MIR` rewrite for the entire `Location` with the `SubLocs` equal to `[]`

Reply:
We generally treat statements and terminators the same in `unlower` and `distribute`.  For `CellSet` in particular, it looks like `mir_op` does emit the rewrite on the statement itself (empty `Vec<SubLoc>`).

Reply:
That's better